### PR TITLE
The Return of Spray Paint

### PR DESCRIPTION
**Pull request by DEATHB4DEFEAT**
# Description

Fixes some issues described in https://github.com/space-wizards/space-station-14/issues/26257 and does some other things.
This PR's changes are licensed under MIT.

---

# Changelog

:cl:
- add: Added spray-painting back

### CODE REVIEW

File: Content.Server/Paint/PaintSystem.cs [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+                            continue;
+
+                        EnsureComp<PaintedComponent>(slotEnt.Value, out var slotToPaint);
+                        EnsureComp<AppearanceComponent>(slotEnt.Value);
+                        slotToPaint.Color = entity.Comp.Color;
+                        _appearanceSystem.SetData(slotEnt.Value, PaintVisuals.Painted, true);
+                        Dirty(slotEnt.Value, slotToPaint);
+                    }
+                }
+            }
```
Comment (VMSolidus) [inline Content.Server/Paint/PaintSystem.cs:150]:
```suggestion
            if (HasComp<InventoryComponent>(target) // Paint any clothing the target is wearing.
                && _inventory.TryGetSlots(target, out var slotDefinitions))
                foreach (var slot in slotDefinitions)
                {
                    if (!_inventory.TryGetSlotEntity(target, slot.Name, out var slotEnt)
                        || HasComp<PaintedComponent>(slotEnt.Value)
                        || !entity.Comp.Blacklist?.IsValid(slotEnt.Value, EntityManager) != true
                        || HasComp<RandomSpriteComponent>(slotEnt.Value)
                        || HasComp<HumanoidAppearanceComponent>(slotEnt.Value))
                        continue;

                    EnsureComp<PaintedComponent>(slotEnt.Value, out var slotToPaint);
                    EnsureComp<AppearanceComponent>(slotEnt.Value);
                    slotToPaint.Color = entity.Comp.Color;
                    _appearanceSystem.SetData(slotEnt.Value, PaintVisuals.Painted, true);
                    Dirty(slotEnt.Value, slotToPaint);
                }
```
Gaming.

File: Content.Server/Paint/PaintSystem.cs [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    }
+
+
+    private void OnInteract(EntityUid uid, PaintComponent component, AfterInteractEvent args)
+    {
+        if (!args.CanReach)
+            return;
+
+        if (args.Target is not { Valid: true } target)
+            return;
```
Comment (VMSolidus) [inline Content.Server/Paint/PaintSystem.cs:51]:
```suggestion
        if (!args.CanReach
            || args.Target is not { Valid: true } target)
            return;
```

File: Content.Server/Paint/PaintSystem.cs [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        var doAfterEventArgs = new DoAfterArgs(EntityManager, user, component.Delay, new PaintDoAfterEvent(), uid, target: target, used: uid)
+        {
+            BreakOnUserMove = true,
+            BreakOnTargetMove = true,
+            NeedHand = true,
+            BreakOnHandChange = true,
+        };
+
+        _doAfterSystem.TryStartDoAfter(doAfterEventArgs);
+    }
```
Comment (VMSolidus) [inline Content.Server/Paint/PaintSystem.cs:88]:
```suggestion
    private void PrepPaint(EntityUid uid, PaintComponent component, EntityUid target, EntityUid user) =>
        _doAfterSystem.TryStartDoAfter(new DoAfterArgs(EntityManager, user, component.Delay, new PaintDoAfterEvent(), uid, target: target, used: uid)
        {
            BreakOnUserMove = true,
            BreakOnTargetMove = true,
            NeedHand = true,
            BreakOnHandChange = true,
        });
```
You don't have to do this particular one, I was just curious of it was possible. :3

File: Content.Server/Paint/PaintSystem.cs [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+                        continue;
+
+                    EnsureComp<PaintedComponent>(slotEnt.Value, out var slotToPaint);
+                    EnsureComp<AppearanceComponent>(slotEnt.Value);
+                    slotToPaint.Color = color;
+                    _appearanceSystem.SetData(slotEnt.Value, PaintVisuals.Painted, true);
+                    Dirty(slotEnt.Value, slotToPaint);
+                }
+            }
+        }
```
Comment (VMSolidus) [inline Content.Server/Paint/PaintSystem.cs:190]:
```suggestion
        if (HasComp<InventoryComponent>(target)
            && _inventory.TryGetSlots(target, out var slotDefinitions))
            foreach (var slot in slotDefinitions)
            {
                if (!_inventory.TryGetSlotEntity(target, slot.Name, out var slotEnt)
                    || blacklist.IsValid(slotEnt.Value, EntityManager))
                    continue;

                EnsureComp<PaintedComponent>(slotEnt.Value, out var slotToPaint);
                EnsureComp<AppearanceComponent>(slotEnt.Value);
                slotToPaint.Color = color;
                _appearanceSystem.SetData(slotEnt.Value, PaintVisuals.Painted, true);
                Dirty(slotEnt.Value, slotToPaint);
            }
```

File: Content.Server/Paint/PaintSystem.cs [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+        {
+            var quantity = solution.RemoveReagent(reagent.Comp.Reagent, reagent.Comp.ConsumptionUnit);
+            if (quantity > 0)// checks quantity of solution is more than 0.
+                return true;
+
+            if (quantity < 1)
+                return false;
+        }
+        return false;
+    }
```
Comment (VMSolidus) [inline Content.Server/Paint/PaintSystem.cs:211]:
```suggestion
    private bool CanPaint(Entity<PaintComponent> reagent, EntityUid target)
    {
        if (HasComp<HumanoidAppearanceComponent>(target) 
            || HasComp<SubFloorHideComponent>(target)
            || !_solutionContainer.TryGetSolution(reagent.Owner, reagent.Comp.Solution, out _, out var solution))
            return false;

        var quantity = solution.RemoveReagent(reagent.Comp.Reagent, reagent.Comp.ConsumptionUnit);
        return (quantity > 0);
    }
```

File: Content.Client/Paint/PaintVisualizerSystem.cs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    protected override void OnAppearanceChange(EntityUid uid, PaintedComponent component, ref AppearanceChangeEvent args)
+    {
+        var shader = _protoMan.Index<ShaderPrototype>(component.ShaderName).Instance();
+
+        if (args.Sprite == null)
+            return;
+
+        // What is this even doing? It's not even checking what the value is.
+        if (!_appearance.TryGetData(uid, PaintVisuals.Painted, out bool isPainted))
+            return;
```
Comment (VMSolidus) [inline Content.Client/Paint/PaintVisualizerSystem.cs:36]:
```suggestion
        var shader = _protoMan.Index<ShaderPrototype>(component.ShaderName).Instance();

        if (args.Sprite == null)
            return;

        // What is this even doing? It's not even checking what the value is.
        if (!_appearance.TryGetData(uid, PaintVisuals.Painted, out bool isPainted))
            return;
```
I'm reasonably sure that _protoMan has a TryIndex that can be used as a guard clause here, and thus all 3 of these can be combined.

**Review by VMSolidus — CHANGES_REQUESTED** [review]

**Comment by DEATHB4DEFEAT** [comment]
Ignore the code style stuff in here; https://github.com/Simple-Station/Einstein-Engines/pull/1226

**Review by VMSolidus — APPROVED** [review]

**Pull request merged**